### ISSUE TITLE
Merge Editor: Use existing merge conflict theme colors?

### ISSUE THREAD

**Issue by daviddossett** [microsoft/vscode#158797]
It occured to me that many users have familiarity with the inline merge conflict theme colors either in the default themes or in others. We could consider using them to indicate incoming/current changes in the merge editor too. The goal being to slightly better bridge the gap betwen the two merging options.

We have another layer of handled/unhandled but I'm sure that could be handled too. We could even simply just use those repspective colors once the change has been handled rather than the generic color used today.

<img width="629" alt="CleanShot 2022-08-22 at 11 00 26@2x" src="https://user-images.githubusercontent.com/25163139/185988191-c52d1185-5292-485f-ad87-421ae961bceb.png">

<img width="931" alt="CleanShot 2022-08-22 at 10 54 45@2x" src="https://user-images.githubusercontent.com/25163139/185988197-24458d30-e568-4516-9828-966ba49cf55d.png">

**Comment by bpasero** [comment]
I am not sure if it is related, but for me I spend 99.99% of my time looking at the diff editor and understanding changes and maybe 0.01% of my time looking at merge conflicts. So if we talk about aligning colors, please consider the colors we already use for added, modified, changed lines:

![image](https://user-images.githubusercontent.com/900690/186608910-51df2900-81bd-4703-bfb9-b23410404759.png)

I think that would make me feel at home in the merge editor easily.

**Comment by daviddossett** [comment]
Matched the default diff background color and word diff colors as a start which I think is close to what @hediet originally had with some minor tweaks in #159216:

![CleanShot 2022-08-25 at 11 26 32@2x](https://user-images.githubusercontent.com/25163139/186741099-8ecc7845-fb20-4d59-a678-00ee67e697b3.png)

**Comment by rzhao271** [comment]
I had to resolve a conversation for auto-merge to kick in, so the fix is currently in for September. The issue needs to be made into a candidate if we need this for Stable.

**Comment by hediet** [comment]
@daviddossett I think a candidate makes sense here?

**Comment by daviddossett** [comment]
Reopening to explore this after reading more feedback. This could be a way to make the jump from inline decorations to the merge editor less jarring. Here's a simple example. 

<img width="701" alt="CleanShot 2022-09-07 at 20 17 22@2x" src="https://user-images.githubusercontent.com/25163139/189026531-4f599f27-ae21-407f-9326-133cf0481a46.png">

cc @isidorn

**Comment by isidorn** [comment]
@daviddossett makes sense and I fully support it. Use the same color + 1 new colour that means "not reacted on". This new colour could remain the current yellow.

**Comment by daviddossett** [comment]
@hediet  I assume we will need some code changes to distinguish the diff colors by incoming/current? I can make the color changes but I think I could only apply one color to all conflicts at the moment. The only variations I see deal with UI state (focused) or conflict state (handled).

**Comment by daviddossett** [comment]
Explored some other options today based on the latest thinking to use base in the result view when a conflict has not been handled. This does not yet explore how we indicate that the code is from base.

## Remove color in result once handled

This retains the incoming/current colors for the two sides and uses a yellow color to indicate the conflicting line(s) in the result view. It removes the yellow background once the conflict has been handled. Note that the side not taken gets dimmed—another VS Win inspiration ;)

https://user-images.githubusercontent.com/25163139/189238515-9711cce4-0566-424f-a4a6-049300c9c86f.mp4

## Using incoming/current colors for both sides and result

This uses the existing incoming/current colors for the two sides but also represents a handled change using the color in the result view. This could be confusing since the colors represent the conflict, not the resolution. But it could help users understand what the result is comprised of. 

https://user-images.githubusercontent.com/25163139/189238397-876528b8-e1b9-4d4a-9a31-de04e90a2a86.mp4

**Comment by hediet** [comment]
The idea to grey out ignored changes is interesting! We should schedule a meeting and discuss all of your ideas.

**Comment by Mingpan** [comment]
One releated question for @hediet: 

There is an undefined color var in the .../workbench/contrib/mergeEditor/browser/view/media/mergeEditor.css "--vscode-mergeEditor-conflict-unhandledUnfocused-background". All the other colors are defined in ../workbench/contrib/mergeEditor/browser/view/colors.ts except this one.

Is this expected or am I missing something? Is it decided if the conflict area will have a backgound color? If so, will it be possible for themes to define this background color? Thanks!

**Comment by hediet** [comment]
Hey Mingpan!

I think this is probably an oversight. Conflict areas had a background once, but we decided to revert that.

What color theme do you have in mind? How will the conflict background interact with the diff text background color?

**Comment by Mingpan** [comment]
Thanks Henning for the quick response! Is the revert decision fixed, or would you consider changing it at some point?

> What color theme do you have in mind? How will the conflict background interact with the diff text background color?

We use a purple color with 20% alpha for the unhandled conflict background, #dc64dc33 to be precise. 

For text, I believe we use the same purple color but higher alpha (i.e., less transparant) for all text backgrounds, if the text is inside an unhandled conflict area. If it's not a conflict area or the conflict is resolved, it falls back to normal diff text background colors (green / red).

**Comment by hediet** [comment]
Can you share a screenshot?

**Comment by Mingpan** [comment]
Sure. Here's a screenshot of simple conflict area shown in the VS Code's native merge editor (area background #dc64dc33, text background #dc64dc66):

<img width="1728" alt="ConflictAreaColor" src="https://user-images.githubusercontent.com/15076498/212725231-a00c5bed-48f7-4059-afb9-723590b2120f.png">

After accepting either side, it changes to the following normal diff colors, essentially identical to what the merge editor currently shows:

<img width="1728" alt="NoConflictArea" src="https://user-images.githubusercontent.com/15076498/212725451-25d64c32-a364-44c2-b7b8-de5ad334a713.png">

We also haven't fully decided on the exact color & interaction with diff line and text, though.

**Comment by hediet** [comment]
I see. If you like, you can come up with a PR to add back the background color! We won't set it by default in our themes, but some experimenting in external themes doesn't hurt.
